Add source type hierarchy

charleskawczynski opened the pull request:
# Description

This is a first step towards fixing #395, #1527, #1036, and #1633.

It seems that all sources could be written this way. I think this has a few advantages:
 - Users can overload `eq_tends` and add / remove terms in a hierarchical fashion, while not affecting the other flux terms
 - `flux` and `source` can be used by diagnostics to export/monitor the budget.
 - This could also be used for specifying time-explicit/implicit source terms more easily via types
 - We can remove the generated functions in `Atmos/Model/source.jl`

This is just a prototype for now. I figured I'd open this PR to get feedback. This seems to work fine for `experiments/AtmosLES/bomex_single_stack.jl` (both on CPU and GPU)

One potential concern is, can we continue to unroll the tendencies if the number of sources gets large. @vchuravy? I can't imagine we'll reach over 100 (ever), but I could imagine reaching/needing granularity of, say 20.

<!--- Please fill out the following section --->

I have

- [x] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [x] Updated the documentation to reflect changes from this PR.

<!--- Please leave the following section --->

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

kpamnany commented:
This looks like a good way to break things down, but not sure I have useful input.

charleskawczynski commented:
bors try

thomasgibson commented:
This is a great start! I will spend some time reading over this carefully --- this is closely related to questions many of us have about how we abstract the various terms in Clima.

thomasgibson reviewed (commented):
So this PR basically captures two main things: (1) an abstraction for tendencies (which I think we should explicitly define in the documentation, e.g. tendency vs term); and (2) sums of tendencies (not as an explicitly types object, but rather as a wrapper).

I think helps simplify how the AtmosModel (and similar subtypes of BalanceLaw) specifies the specification of fluxes. I am in favor of proceeding in this fashion towards something which exposes the tendencies in a more clear way.

My only concern is how this can be leveraged by the rest of clima. Right now, these changes are mostly cosmetic. But does this typing help us factorize out specific terms in a given balance law? I would be very interested to explore this more with you @charleskawczynski --- as I think this helps us get closer to what we really want at the higher-level interfaces

thomasgibson commented on src/Atmos/Model/tendencies.jl:
```diff
@@ -0,0 +1,15 @@
+#####
+##### Tendency specification
+#####
```
We should be very clear what we define as a tendency --- I think there is a lot of confusion between the various labels we commonly use (tendency, term, etc)

charleskawczynski replied:
I agree that we should have a clear and documented terminology. I've been using "tendency" to mean any RHS term, be it a flux or non-conservative source, in a spatially/temporally-continuous sense.

valeriabarra replied:
This would be super helpful for me, too - as I have been a bit puzzled by this terminology to be honest, but I am not familiar with other climate codes, so I thought that it was due to my unfamiliarity

charleskawczynski replied:
IIUC, tendencies of `Y_j` are individual terms on the RHS of the time-stepped governing equation for `Y_j`, which have units `units(Y_j)/time`, i.e., `∂_t Y_j = Σ_i tendency_i`. Maybe others can chime in? I believe it's a temporally-continuous concept, so there shouldn't be ambiguities regarding the implicitness of the way each tendency is treated.

charleskawczynski replied:
If people prefer "term", I think that would be fine, but I think "tendency" is actually more specific. It not only restricts the units of the quantity, but also the context (a single term in a time-stepped equation). I was a bit skeptical of this terminology at first (we usually use RHS terms in Mech Eng), but I've come around to like it.

charleskawczynski replied:
Perhaps there's ambiguity with the time-dependent term and sign, however. I realize the definition I gave is vague to disambiguate `∂_t Y_j = Σ_i tendency_i` from `Σ_i tendency_{i,j} = 0`. So, we'd need a sign convention with this definition.

charleskawczynski replied:
I'll move the `tendencies_sum.jl` to `BalanceLaws/` and add some documentation there.

blallen replied:
How about both? `tendency term`? `tendency` can refer to both a `tendency_i` and `∂_t Y_j`, which might be ambiguous, and `term` can refer to any type of term in an equation. combining the two seems to be specific enough

valeriabarra replied:
Thanks @charleskawczynski for the explanation. So from what I gather, these are the spatial terms, as in everything else that equals the time derivative, as you wrote above `∂_t Y_j = Σ_i tendency_i` . But you also specified RHS terms, meaning when you use an explicit time stepper. What about when you use an implicit one? Are the spatial terms still called tendencies?

charleskawczynski replied:
I apologize, I've abused the phrase "RHS term" a bit. My understanding is that tendencies are temporally-continuous. So, whether a term is treated explicitly / implicitly is orthogonal. I.e., we could label any tendency to be treated explicitly / implicitly in time.

valeriabarra replied:
Okay, so basically I gather that is the totality of all spatial terms that equal the time variation in a conservation form, does not matter if/how discretized

blallen commented:
@charleskawczynski do you think you can mock up an example where you use this structure to actually pick out one of the terms? What you have looks good, but I feel like the real test is if adding all this structure actually lets you do the selection you want

charleskawczynski commented:
> So this PR basically captures two main things: (1) an abstraction for tendencies (which I think we should explicitly define in the documentation, e.g. tendency vs term); and (2) sums of tendencies (not as an explicitly types object, but rather as a wrapper).

Agreed

> I think helps simplify how the AtmosModel (and similar subtypes of BalanceLaw) specifies the specification of fluxes. I am in favor of proceeding in this fashion towards something which exposes the tendencies in a more clear way.

🎉 

> My only concern is how this can be leveraged by the rest of clima.

How so? This same pattern can be applied to any balance law, AFAICT.

> But does this typing help us factorize out specific terms in a given balance law? I would be very interested to explore this more with you @charleskawczynski --- as I think this helps us get closer to what we really want at the higher-level interfaces!

Perhaps I'm unclear about what is needed for the multirate bits. I imagine that this level of granularity would make specifying balance laws easier via types/tuples.

charleskawczynski commented:
> @charleskawczynski do you think you can mock up an example where you use this structure to actually pick out one of the terms? What you have looks good, but I feel like the real test is if adding all this structure actually lets you do the selection you want

I can try to mock something up, is there an example that I can look at? Also, I have some ideas that might help trim some code.

charleskawczynski commented:
bors try

charleskawczynski commented:
bors try

blallen commented:
> > @charleskawczynski do you think you can mock up an example where you use this structure to actually pick out one of the terms? What you have looks good, but I feel like the real test is if adding all this structure actually lets you do the selection you want
> 
> I can try to mock something up, is there an example that I can look at? Also, I've not trimmed out everything that can be removed to help make the diff more readable.

I mean actually finding something like 'mass tendency' from a constructed `AtmosModel` struct. Just have a short example where you initialize an `AtmosModel` and then check the mass tendency somehow

charleskawczynski commented:
> I mean actually finding something like 'mass tendency' from a constructed `AtmosModel` struct. Just have a short example where you initialize an `AtmosModel` and then check the mass tendency somehow

Meaning just print the tendencies for a given atmos model? I was thinking we could add and call something like
```julia
function show_atmos_tendencies(m)
  T = (Flux1ˢᵗOrder(), Flux2ⁿᵈOrder(), NonConservative())
  @show ρ_tend.(Ref(m), T)
  @show ρu_tend.(Ref(m), T)
  @show ρe_tend.(Ref(m), T)
  @show ρq_tot_tend.(Ref(m), T)
end
```
in the driver.

charleskawczynski commented:
bors try

blallen commented:
> > I mean actually finding something like 'mass tendency' from a constructed `AtmosModel` struct. Just have a short example where you initialize an `AtmosModel` and then check the mass tendency somehow
> 
> Meaning just print the tendencies for a given atmos model? I was thinking we could add and call something like 
> 
> ```julia
> function show_atmos_tendencies(m)
>   T = (Flux1ˢᵗOrder(), Flux2ⁿᵈOrder(), NonConservative())
>   @show ρ_tend.(Ref(m), T)
>   @show ρu_tend.(Ref(m), T)
>   @show ρe_tend.(Ref(m), T)
>   m.moisture isa EquilMoist || m.moisture isa NonEquilMoist && @show ρq_tot_tend.(Ref(m), T)
> end
> ```
> 
> in the driver.

Yep, something like this exactly. Would this show them split into 1st, 2nd, and source terms? Also is it possible to pull out just `ρeAdvect` or `ρePressure`?

charleskawczynski commented:
> Yep, something like this exactly.

Will do.

> Would this show them split into 1st, 2nd, and source terms?

Yep, each of the 3 lines should print a tuple of tuples-- the outer being one of `Flux1ˢᵗOrder(), Flux2ⁿᵈOrder(), NonConservative()`, and the inner being the particular tendency.

> Also is it possible to pull out just `ρeAdvect` or `ρePressure`?

What exactly do you mean by pull out? I think we should be able to remove types from the tuple (using `filter / isa` and potentially abstract types), and identify if a particular type exists (just with `isa`).

blallen commented:
> > Yep, something like this exactly.
> 
> Will do.
> 
> > Would this show them split into 1st, 2nd, and source terms?
> 
> Yep, each of the 3 lines should print a tuple of tuples-- the outer being one of `Flux1ˢᵗOrder(), Flux2ⁿᵈOrder(), NonConservative()`, and the inner being the particular tendency.
> 
> > Also is it possible to pull out just `ρeAdvect` or `ρePressure`?
> 
> What exactly do you mean by pull out? I think we should be able to remove types from the tuple (using `filter / isa` and potentially abstract types), and identify if a particular type exists (just with `isa`).

Ah, will the `Flux1ˢᵗOrder()` tuple contain `ρeAdvect` and `ρePressure`? And for the latter question, I mean something like `energy_advect = return_individual_tendency(model, ρeAdvect)` and it returns how this term is specified (or the value at a nodal point, or an average or something if you want to be fancy for diagnostics).

charleskawczynski commented:
> Ah, will the `Flux1ˢᵗOrder()` tuple contain `ρeAdvect` and `ρePressure`? And for the latter question, I mean something like `energy_advect = return_individual_tendency(model, ρeAdvect)` and it returns how this term is specified (or the value at a nodal point, or an average or something if you want to be fancy for diagnostics).

Yes, it will contain a tuple of both. Oh, the latter point would be cool! I think we could do something like that if, for each source term, we include a doc string of the precise term (perhaps in LaTeX/Unicode). Here's what it currently shows (for `tutorials/Atmos/heldsuarez.jl` and not all source terms added):
```julia
┌ Info: Tendency list:
│     ρ (Flux1ˢᵗOrder) = (ρAdvect())
│     ρ (NonConservative) = ()
│     ρu (Flux1ˢᵗOrder) = (ρuAdvect(), ρuHydroPress())
│     ρu (NonConservative) = ()
│     ρe (Flux1ˢᵗOrder) = (ρeAdvect(), ρePressure())
└     ρe (NonConservative) = ()
```

charleskawczynski commented:
bors try

kmdeck commented:
> Yep, that list looks great. And in addition to the string documentation, I think it would be nice if you could use these types to return the nodal values for diagnostics, but that's probably beyond the scope of this PR.

hey Charlie, if I am understanding the general point, this lets us treat the RHS in a more granular way. it looks like this code makes it so all RHS terms are treated like we treat sources now. so we would need to define e.g. a type `SecondOrderHeatFlux` for the sole RHS term in the simple heat equation, and then we have a method of a function `_flux_` which evaluates this, and eventually this gets added to the sum of tendencies for second order fluxes, which then gets added to the total RHS sum. 

If down the line different RHS terms get treated differently (for ODE advancement, eg.), it makes sense that they each have a type that other functions can dispatch on.

More granularity is helpful for debugging too - I'm not sure if we can do this currently with sources, or if this is what Brandon is getting at with his comment above, but it would be helpful to be able to call these functions in a callback, i.e. to get a particular RHS term evaluated every X time steps. 

my personal preference would also be to call them e.g. FluxFirstOrder and not Flux1^{st}Order :)

blallen commented:
> my personal preference would also be to call them e.g. FluxFirstOrder and not Flux1^{st}Order :)

yes please

charleskawczynski commented:
> hey Charlie, if I am understanding the general point, this lets us treat the RHS in a more granular way. it looks like this code makes it so all RHS terms are treated like we treat sources now. so we would need to define e.g. a type `SecondOrderHeatFlux` for the sole RHS term in the simple heat equation, and then we have a method of a function `_flux_` which evaluates this, and eventually this gets added to the sum of tendencies for second order fluxes, which then gets added to the total RHS sum.

Yes, exactly. That said, we don't necessarily need to use this pattern everywhere (e.g., tutorials), since it's compatible with the existing interface. But I imagine that this will be useful in the Atmos/Ocean/Land/Ice models, since we'll likely have a long tendency list per model.

> If down the line different RHS terms get treated differently (for ODE advancement, eg.), it makes sense that they each have a type that other functions can dispatch on.

Yes, this should provide an alternative way to hook specific terms based on the tendency types.

> More granularity is helpful for debugging too - I'm not sure if we can do this currently with sources, or if this is what Brandon is getting at with his comment above, but it would be helpful to be able to call these functions in a callback, i.e. to get a particular RHS term evaluated every X time steps.

yes, I believe that's the idea behind optionally exporting a budget, which should be possible in an automated way using this tendency specification.

> my personal preference would also be to call them e.g. FluxFirstOrder and not Flux1^{st}Order :)

Agreed, I was unsure what to use for some names.

charleskawczynski commented:
bors try

charleskawczynski commented:
I spoke a bit with @kpamnany about adding a callback somewhere to print the tendency names/values, but I'm actually not sure which test/driver to use. I'll drop what I have here for now:

```julia
function tend_tuple(var_tend, m, args_f1, args_f2, args_s)
    return (
    ((Atmos.format_tend(tend),Atmos._flux_(tend, args_f1...)) for tend in var_tend(m, Flux1ˢᵗOrder()))...,
    ((Atmos.format_tend(tend),Atmos._flux_(tend, args_f2...)) for tend in var_tend(m, Flux2ⁿᵈOrder()))...,
    ((Atmos.format_tend(tend),Atmos._source_(tend, args_s...)) for tend in var_tend(m, NonConservative()))...,
    )
end

tend_cb = GenericCallbacks.EveryXSimulationTime(every_x_simulation_time) do
    dg = solver_config.dg
    m = dg.balance_law
    direction = dg.direction
    vs_s = vars_state(m, Prognostic(), eltype(dg.Q))
    vs_a = vars_state(m, Auxiliary(), eltype(dg.Q))
    vs_d = vars_state(m, GradientFlux(), eltype(dg.Q))
    vs_hd = vars_state(m, Hyperdiffusive(), eltype(dg.Q))
    loc = (1,:,1)                                                                                        # print tendencies at a specific point in space
    state = Vars{vs_s}(dg.Q[loc...])
    aux = Vars{vs_a}(dg.state_auxiliary[loc...])
    diffusive = Vars{vs_d}(dg.state_gradient_flux[loc...)
    hyperdiff = Vars{vs_hd}(dg.states_higher_order[2][loc...])

    args_f1 = (m, state, aux, t, ts)
    args_f2 = (m, state, aux, t, ts, diffusive, hyperdiff)
    args_s = (m, state, aux, t, ts, direction)

    ρ_tends = tend_tuple(Atmos.ρ_tend, m, args_f1, args_f2, args_s)
    ρu_tends = tend_tuple(Atmos.ρu_tend, m, args_f1, args_f2, args_s)
    ρe_tends = tend_tuple(Atmos.ρe_tend, m, args_f1, args_f2, args_s)
    @show ρ_tends
    @show ρu_tends
    @show ρe_tends

    if m.moisture isa EquilMoist || m.moisture isa NonEquilMoist
        ρq_tot_tends = tend_tuple(Atmos.ρq_tot_tend, m, args_f1, args_f2, args_s)
        @show ρq_tot_tends
    end

    nothing
end;
```

kpamnany commented:
> I spoke a bit with @kpamnany about adding a callback somewhere to print the tendency names/values

This will go into a diagnostics group so the callback is really an example for developing that rather than an example for use by an experiment.

charleskawczynski commented:
bors try

charleskawczynski commented:
bors try

charleskawczynski commented:
Alright, I
 - Renamed `Flux1ˢᵗOrder` to `FluxFirstOrder` and `Flux2ⁿᵈOrder` to `FluxSecondOrder`
 - Renamed the existing `Atmos.Source` to `Atmos.AbstractSource` (which will eventually be phased out) and `BalanceLaws.NonConservative` to `BalanceLaws.Source`
 - Moved `tendencies_sum.jl` to `BalanceLaws/` and realized that, for the interface to work with the Atmos/Ocean/Land, we'll likely want the tendencies sum to be pretty generic, so I just changed the sum to slurp/splat all args after the tendency type.
 - After moving/generalizing the interface, I realized that `Σfluxes` and `Σsources` had exactly the same signature, and just combined it to `Σtendencies`, which simply calls `tendency`.

I'm pleased with how things have evolved. I think the only thing left is to do a final pass on documentation. @thomasgibson @valeriabarra @blallen do you have thoughts on improvements?

charleskawczynski commented:
bors try

simonbyrne commented:
How about something like this

```julia
abstract type PrognosticVariable end
struct Mass <: PrognosticVariable end
struct Momentum <: PrognosticVariable end
struct Energy <: PrognosticVariable end

# something you can  "add" to the RHS of an equation set that affects one or more variables
abstract type Tendency end
struct Advection <: Tendency end
struct PressureGradient <: Tendency end
struct Subsidence <: Tendency end
struct Total{TS} <: Tendency
    tends::TS
end

# something that we need to compute
abstract type Term end
struct FluxFirstOrder{V<:PrognosticVariable,T<:Tendency,R<:Rate} <: Term
    var::V
    tend::T
end
struct FluxSecondOrder{V<:PrognosticVariable,T<:Tendency} <: Term end
struct Source{V<:PrognosticVariable,T<:Tendency} <: Term end

# e.g.
compute(::FluxFirstOrder{Advection,Mass}, state) = state.ρu
compute(::FluxFirstOrder{Advection,Momentum}, state) = state.ρu ⊗ (state.ρu ./ state.ρ)
compute(::FluxFirstOrder{Advection,Energy}, state) = (state.ρu ./ state.ρ) .* state.ρe


function compute(::Source{Subsidence,Mass}, state)
    z = altitude(m, aux)
    w_sub = subsidence_velocity(s.subsidence, z)
    k̂ = vertical_unit_vector(m, aux)
    return -state.ρ * w_sub * dot(k̂, diffusive.moisture.∇q_tot)
end
function compute(::Source{Subsidence,Energy}, state)
    z = altitude(m, aux)
    w_sub = subsidence_velocity(s.subsidence, z)
    k̂ = vertical_unit_vector(m, aux)
    return -state.ρ * w_sub * dot(k̂, diffusive.∇h_tot)
end

# could define extra terms to force common subexpression elimination?
struct SubsidenceVelocity <: Term end
compute(::SubsidenceVelocity, state) = ...

# sum up all tendencies for a given variable
compute(f1::FluxFirstOrder{V,Total{TS}}, args...) where {V,TS} =
    sum(tend -> compute(FluxFirstOrder(f1.tend, term),args...), f1.tend.tends)
FluxFirsrOrder(var::Momentum) = FluxFirstOrder(var,Total((Advection,PressureGradient)))
```

For multirate schemes, we could either add an extra parameter to each `FluxFirstOrder`, `FluxSecondOrder`, etc, or we could wrap each `Tendency`
```julia
struct Rate{R, T<: Tendency} <: Tendency
    rate::R
    tend::T
end

compute(::FluxFirstOrder{Rate{Linear, Advection},Mass}, state) = state.ρu
```

simonbyrne commented:
Actually it occurs to me that in the above, `PrognosticVariable` could just be a `Term` as well.

simonbyrne commented on src/Atmos/Model/tendencies_multiphysics.jl:
```diff
@@ -0,0 +1,20 @@
+subsidence_velocity(subsidence::Subsidence{PV, FT}, z::FT) where {PV, FT} =
+    -subsidence.D * z
+
+struct Pressure{PV} <: FluxFirstOrderTendency{PV} end
```
`Pressure` is a quantity so I don't think we should call the tendency that as well. How about:
```suggestion
struct PressureGradient{PV} <: FluxFirstOrderTendency{PV} end
```
since it is the tendency contribution of the gradient of the pressure.

charleskawczynski replied:
@simonbyrne I'm realizing that `PressureGradient` may only be appropriate for the momentum equation, whereas the pressure term for the energy equation has no gradients (it's a scalar equation). Should we just leave this as `Pressure`? @ilopezgp maybe you can weigh in? Alternatively, we could leave `PressureGradient{PV<:Momentum}` and just add `Pressure{Energy}`. I have no strong opinions on this, so long the names are appropriate.

charleskawczynski replied:
For now, I've added `Pressure` and `PressureGradient` separately, but we can combine them if it's preferred--it's a simple switch.

simonbyrne commented on src/Atmos/Model/tendencies_energy.jl:
```diff
@@ -0,0 +1,33 @@
+##### First order fluxes
+#####
+
+function tendency(::Advect{Energy}, m, state, aux, t, ts, direction)
```
I don't think we should use `tendency` here, since it isn't applied directly as a tendency (its divergence is). How about:
```suggestion
function flux(::Advect{Energy}, m, state, aux, t, ts, direction)
```

charleskawczynski commented:
Ok, I've renamed a bunch of things and added some docs to hopefully clarify precisely what each newly added piece is/does. Most of these docs were added to `src/BalanceLaws/tendency_types.jl`. To avoid terminology confusion `tendency` was split back into `flux` and `source`. The other recent major change was introducing an `AbstractPrognosticVariable`, which allows us to basically dispatch on the rows (equations) and columns (flux^1, flux^2, source) for each of the methods.

I think I'm happy with the way things are looking now. I'm going to look around and try to add a few more flux tendencies and source tendencies (TM 😄) that may cover important edge cases that I don't want to overlook. Then, I'll be happy to merge this thing, woohoo! 🎉

charleskawczynski commented:
bors try

charleskawczynski commented:
bors try

charleskawczynski commented:
Alright, I made some changes that I think simplify some things that were previously bugging me. I think we're ready to go!

charleskawczynski commented:
bors r+

charleskawczynski commented:
bors r+
